JBPM-5409: Third bug fixing iteration and added more tests.

romartin opened the pull request:
Hey @manstis ,
Here are some bug fixings, improvments and added more unit tests as well. 
See last comment on [this ticket](https://issues.jboss.org/browse/JBPM-5409)
Main updates are:
- Added property meta-types
The idea for this is supporting the resize, change name from widgets and any other further needs to bind ui properties or properties used in some Stunner components  with the model, by using the adapters . This replaces the @NameProperty and try to make properties meta-data more generic. This approach was the faster I coulld find for now, just by changing a bit the supported annotation and the annot processors.

@manstis  Please -  this includes [the first commit of this PR you already have in your branch](https://github.com/manstis/kie-wb-common/pull/1/commits),you'll have to merge this,, sorry.  And plz double check as weill that this commit does not broke your CM bean generation stuff. Lemme know if any issues! thx!

- Added RequestCommandManager
As already discussed too, this is a command manager which all comands that are executed on same "client request" ( for now, from mouseDown until mouseUp ) are further undo as a single composite one, so different components that should execute commands can inject either the session or the request command manager.

@hasys : Sorry had this PR before the code style refactors, will apply from now on! :)

Thanks!

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/command/RequestCommandManager.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    /**
+     * Starts a new client request.
```
Oh no it doesn't - copy and paste error.

romartin replied:
opops, sorry will review those javadocs

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/command/impl/CommandRegistryListener.java:
```diff
@@ -0,0 +1,52 @@
+    protected abstract CommandRegistry<Command<C, V>> getRegistry();
+
+    @Override
+    public void onAllow( final C context,
```
Should this have an empty default implementation; or is it for sub-classes to **always** implement?

romartin replied:
MM there are two implmentation for this that are inner classes, one inside SessionCommandManagerImpl and another instide RequestCommandManager. Currently the registry instance provided by the abstract method is the one from current "canvas session", so it's better just on the abstract method implementation obtain the registry instance at runtime for current session. But yes other components that always use same registry could just a default implementation for this, I can create it np. Is this what you mean? are we on same page? :)

manstis replied:
I don't think we're on the same page. This method, in the **abstract** class, is empty. Does nothing. Yet you say the two concrete implementations of this class provide their own implementation of this method. So why have this empty default implementation at all!? It is an **abstract** class so doesn't need it.

romartin replied:
ahhhhh sorry you talk about the `onAllow` method! I though was about the `getRegistry` one...  so sure, makes sense! :)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/command/impl/CompositeCommandImpl.java:
```diff
@@ -48,22 +55,50 @@
+         * The undo for this composite command will be done by using a reverse order.
+         */
+        public CompositeCommandBuilder<T, V> reverse() {
+            this.reverse = true;
```
Is it best to have this flag; or simply sort the ```List```?

manstis replied:
OK, I see. If we sorted the collection new entries could potentially be at the "wrong end".

romartin replied:
yep

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/command/impl/CompositeCommandImpl.java:
```diff
@@ -48,22 +55,50 @@
+         * The undo for this composite command will be done by using the same insertion order.
+         */
+        public CompositeCommandBuilder<T, V> forward() {
+            this.reverse = false;
```
Is it best to have this flag; or simply sort the ```List```?

romartin commented:
@manstis thanks for the review! I
In general I'm agree on your comment - that are quite some missing TODOs and missing cleans, I wanted to push this as soon as possible to make it available for your work on CM, but yeah you're right, let's improve it, so if you don't mind, I'll provide a new commit on next days...

romartin commented:
@manstis btw what you mean by replace use of `null` by `Optional?`sorry not following you

romartin commented:
@manstis let you know I'm still working on this, when I find some time. So please be patient... will rebase it with latest master and try to add more fixes and implement as much the missing TOODs... 
If you get the point where you need to use improve of fix the commands API and this stuff or need serialization for shape bounds/size you will need this merged, so please lemme know if you get to this point before I have added the new commit so will give it priority. Thanks!

romartin commented:
Hey @manstis ,

I've added a new commit on this PR that fixes your comments above and in addition to the updates on the comments above it contains more fixes and improvments, please see last comment on [this ticket](https://issues.jboss.org/browse/JBPM-5409).

Cannot test neither the project-showcase neither the kie/drools-wb due to [the errai bug](https://issues.jboss.org/browse/ERRAI-1016) - They were working fine some days ago but tbh cannot run the build due to that isuue - but at least could you please check that your CM is not broken with this new commit?

It includes code style and license refactors, and more unit testing.

Thanks!

hasys commented:
Hi @romartin, @manstis,

this bug is fixed now. Try "mvn clean install **-U**" it helped to me.

romartin commented:
oh great thanks for the info @hasys !!! will give it a try with -U ;)

romartin commented:
@wmedvede this includes the fix for kie/drools-wb about the datamodeller docks.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/menu/dev/ShapeSetsMenuItemsBuilder.java:
```diff
@@ -0,0 +1,100 @@
+                          final Callback callback) {
+        final DropDownMenu menu = new DropDownMenu() {{
+            addStyleName( "pull-right" );
+        }};
```
yes good point, but from my point of view, in those cases like others as data structure definitions with primitives types like maps or lists, there should be no leaking issues as we're creating domain model instances for menus or beans that basically contains primitive types that provide the presentation features, and I think it's more clear this way rather than having instance definitions and add the elements on different lines. But again, I think it's more a personal thing rather than an issue on those cases, we can discuss this and change it if there are good reasons, np! :)

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/wires/WiresUtils.java:
```diff
@@ -26,48 +30,55 @@
+
+    public static boolean isWiresConnector(final ShapeView<?> shapeView) {
+        return shapeView instanceof WiresConnector;
+    }
```
well not sure if could be better, as WiresConnector etc are classes from third party libraries, so adding logic that only stunner needs probably is not an option. In addition, after this call probably there are some instance casts to WiresConnector, so the casting have to be done at some point as well.

romartin commented:
Hey @hasys , @manstis
I have rebased this commit from master and I have UNDONE the code style refactor for this PR, so reduced the insertions/additions from 12K/11K to 5K/4K - hope can be easier to review it now. Will be applying new code style on the next PRs that are smaller.
@hasys : I have fixed some of the comments above and all showcases work for me, can you please try again?
@manstis : can you check this does not broke your CM stuff please? 
I have another PR on top this (smaller as well), so would appreciate if we can try to move forward this ticket which is here for long time... thanks!!! :)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-processors/src/main/java/org/kie/workbench/common/stunner/core/processors/MainProcessor.java:
```diff
@@ -783,23 +766,43 @@ private void processFieldName( TypeElement classElement,
+                    final TypeElement t = ( TypeElement ) ( ( DeclaredType ) fieldReturnType ).asElement();
+                    final String fieldName = variableElement.getSimpleName().toString();
+                    result.put( fieldName, t );
+                    messager.printMessage( Diagnostic.Kind.WARNING, "Discovered meta property value " +
```
Would ```NOTE``` be better than ```WARNING```?

manstis commented:
@hasys I cherry-picked this PR into my local master and did not encounter the issue you reported. I think we should be pragmatic about this PR - since it contains some important changes and there are ~4 other people working against master with new development. I propose merge and submit another PR for **all** code formatting. 

As an aside I notice https://github.com/droolsjbpm/kie-wb-common/pull/621 has merge conflicts and https://github.com/droolsjbpm/kie-wb-common/pull/594 an outstanding conversation with @ederign. Both would need addressing before they can be merged.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-shapes/kie-wb-common-stunner-shapes-client/src/main/java/org/kie/workbench/common/stunner/shapes/client/view/BasicShapeView.java:
```diff
@@ -326,14 +316,22 @@ public T addHandler( final ViewEventType type,
+            if ( ViewEventType.TEXT_OVER.equals( type ) ) {
+                textOverHandlerViewHandler = ( ViewHandler<TextOverEvent> ) eventHandler;
+            }
+            {
```
oops

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/canvas/wires/WiresCanvasView.java:
```diff
@@ -37,33 +36,30 @@ public void init() {
     @Override
     public WiresCanvas.View addShape( final ShapeView<?> shapeView ) {
-        if ( isWiresShape( shapeView ) ) {
+        if ( org.kie.workbench.common.stunner.client.lienzo.canvas.wires.WiresUtils.isWiresShape( shapeView ) ) {
```
oops again

manstis replied:
@romartin You missed this one! (removal of FQCN)

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/session/view/ScreenPanelViewImpl.java:
```diff
@@ -0,0 +1,34 @@
+import com.google.gwt.user.client.ui.IsWidget;
+
+@Dependent
+public class ScreenPanelViewImpl extends FlowPanel implements ScreenPanelView {
```
oh good point! refactoring this thanks!

romartin commented:
@manstis  @hasys thanks guys for the review, applying updates. I'm fixing as well the resize for gateways, as @hasys commented it's not working fine. Will let you know when done again. Thanks!

romartin commented:
@manstis @hasys BTW I think it's perfect what Michael says about doing a PR for just code refactoring!

romartin commented:
hey @manstis @hasys , I have done the above updates commented and added some  missing calls on shape views in order to know if the different types are supported for each one. 
@hasys about the resize issue, the current lienzo stuff does not fit quite well on some situations with what stunner expectes, so I have limited the resize support except for lanes (on bpmn), which results useful. See comment on [ResizeControlImpl](https://github.com/romartin/kie-wb-common/blob/JBPM-5409/kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/resize/ResizeControlImpl.java#L301).
Thanks!

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/canvas/wires/WiresCanvasView.java:
```diff
@@ -37,33 +37,30 @@ public void init() {
     @Override
     public WiresCanvas.View removeShape( final ShapeView<?> shapeView ) {
-        if ( isWiresShape( shapeView ) ) {
+        if ( org.kie.workbench.common.stunner.client.lienzo.canvas.wires.WiresUtils.isWiresShape( shapeView ) ) {
```
@romartin ... and this one

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/canvas/wires/WiresCanvasView.java:
```diff
@@ -37,33 +37,30 @@ public void init() {
             WiresShape wiresShape = ( WiresShape ) shapeView;
             wiresManager.deregister( wiresShape );
-        } else if ( isConnector( shapeView ) ) {
+        } else if ( org.kie.workbench.common.stunner.client.lienzo.canvas.wires.WiresUtils.isWiresShape( shapeView ) ) {
```
@romartin ... and this one

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/session/view/ScreenPanelView.java:
```diff
@@ -14,11 +14,12 @@
+import com.google.gwt.user.client.ui.IsWidget;
 
-public interface ShapeSetsNavigator extends Navigator<ShapeSet> {
+public interface ScreenPanelView extends IsWidget {
```
@romartin OK, so your implementations no longer extend ```Widget``` however your interface still does (it should extend ```org.jboss.errai.common.client.api.IsElement``` and focus on DOM elements rather than Widgets). I am however pragmatic about this (Stunner is hardly "complete"). Perhaps log the need to review use of Widgets against https://issues.jboss.org/browse/JBPM-5409 and revisit before we consider Stunner "complete". WDYT?

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/service/AbstractClientDiagramService.java:
```diff
@@ -57,23 +59,26 @@ public void callback( Path path ) {
-            updateClientMetadata( diagram );
-            callback.onSuccess( diagram );
+        diagramServiceCaller.call( new RemoteCallback<M>() {
+            @Override
```
@romartin This is still not fixed.

manstis commented:
@romartin Re: "resize issue" where you now only show the bottom-right CP in ```ResizeControlImpl```. You might need to do something similar in ```BasicShapeView.showControlPoints()``` since this **could** show ```RESIZE``` control points too (although your current use only shows ```MAGNETS``` when moving connector ends).

romartin commented:
@manstis 
- oh you're right about the `BasicShapeView.showControlPoints()`! I'm bypassing this on the ResizeControlImpl and calling directly the wires view methods, good point! trying to fix this now.
- about the Widgets vs IsElement you're right as well, but the problem is that  is not only this screen now that has to be changed, as I'm using `Widget` everywhere now as views. I'm agree on changing this before the release as well - but there are quite some widget on stunner to refactor, even the forms ones, so what you think if we try to get this PR merged and then provide a new PR only for widget's refactorings in the whole stunner?
Meanwhile fixings other comments, thx for the review!

romartin commented:
Hey @manstis 
- I have fixed your comments above and realized that in addition to using the `BasicShapeView.showControlPoints()` in the ResizeControlImpl I had been able to fix the TODO as well about adding the click handler... so now this control delegates all events and handler registrations to the shape view, as it must be :) 
- About the widgets... as commented, I would prefer doing this refactor in a single commit and all toghether... this PR includes others stuff that you probably need as well, up to you!
Thanks for the review again!
